actors/builtin/miner/miner_state.go

```diff
@@ -360,367 +385,400 @@ func (st *State) ForEachSector(store adt.Store, f func(*SectorOnChainInfo)) erro
+			if err != nil {
+				return false, err
+			}
 
-	st.SectorExpirations, err = arr.Root()
-	return err
+			active, err := bitfield.SubtractBitField(live, partition.Faults)
+			if err != nil {
+				return false, err
+			}
+
+			sectorInfos, err := st.LoadSectorInfos(store, active)
+			if err != nil {
+				return false, err
+			}
+
+			if len(sectorInfos) == 0 {
+				// Nothing to do
+				return false, nil
+			}
+
+			// Note: The expiration stored in the sector info is not altered, but remains the initially-scheduled epoch.
+			err = partition.RescheduleExpirations(store, newEpoch, sectorInfos, ssize, quant)
+			if err != nil {
+				return false, err
+			}
+
+			// Make sure we update the deadline's queue as well.
+			rescheduledPartitions = append(rescheduledPartitions, partIdx)
+			return true, nil
+		},
+		// Update deadline.
+		func(dlIdx uint64, dl *Deadline) (bool, error) {
+			if len(rescheduledPartitions) == 0 {
+				return false, nil
+			}
+
+			// Record partitions that now expire at the new epoch.
+			// Don't _remove_ anything from this queue, that's not safe.
+			if err := dl.AddExpirationPartitions(store, newEpoch, rescheduledPartitions, quant); err != nil {
+				return false, xerrors.Errorf("failed to add partition expirations: %w", err)
+			}
+			return true, nil
+		},
+	)
 }
 
-// Adds sectors numbers to faults and fault epochs.
-func (st *State) AddFaults(store adt.Store, sectorNos *abi.BitField, faultEpoch abi.ChainEpoch) (err error) {
-	empty, err := sectorNos.IsEmpty()
+// Assign new sectors to deadlines.
+func (st *State) AssignSectorsToDeadlines(
+	store adt.Store,
+	currentEpoch abi.ChainEpoch,
+	sectors []*SectorOnChainInfo,
+	partitionSize uint64,
+	sectorSize abi.SectorSize,
+	quant QuantSpec,
+) (PowerPair, error) {
+	// We should definitely sort by sector number.
+	// Should we try to find a deadline with nearby sectors? That's probably really expensive.
+	deadlines, err := st.LoadDeadlines(store)
 	if err != nil {
-		return err
-	}
-	if empty {
-		return nil
+		return NewPowerPairZero(), err
 	}
 
-	{
-		st.Faults, err = bitfield.MergeBitFields(st.Faults, sectorNos)
-		if err != nil {
-			return err
-		}
+	// Sort sectors by number to get better runs in partition bitfields.
+	sort.Slice(sectors, func(i, j int) bool {
+		return sectors[i].SectorNumber < sectors[j].SectorNumber
+	})
 
-		count, err := st.Faults.Count()
-		if err != nil {
-			return err
-		}
-		if count > SectorsMax {
-			return fmt.Errorf("too many faults %d, max %d", count, SectorsMax)
+	// TODO: We should either:
+	// 1. Inline entire deadlines into the Deadlines struct, putting PostSubmissions bitfield behind a CID.
+	//   Pro: Clean.
+	//   Con: Large Deadlines object.
+	// 2. Split deadlines into a header/details section, inlining the headers.
+	//   Pro: Small Deadlines object.
+	//   Con: Walking deadlines is now kind of funky, but not terrible.
+	// 3. Keep a separate arrays for total sectors and live sectors.
+	//   Con: Annoying to update.
```

Comment:
There's a note about inlining in the Deadlines struct. Could you add any still-relevant info there (without TODO, since I don't think we intend to do any of them)

Reply:
Eh, all of this comment can go.